### PR TITLE
drivers: rtc: use spinlocks in RTC emul device

### LINKED ISSUE #59901: tests: drivers: rtc: multiple minor issues in rtc_emul mock/fake device
**Describe the bug**

The RTC tests and test mock (emulation) device have multiple minor issues:
* ~~The RTC tests that run in the CI/CD platform do not cover the alarm, update and calibration parts of the API by default although such tests exist in the tests tree.~~
* ~~The RTC test suite does not build on 64 bit architectures.~~
* ~~The RTC emulation device is enabled by default on native_posix for all tests, although it's only required for a single driver.~~
* The RTC driver tests that ran in the CI/CD pipeline are instantiated on all platforms (to be filtered then via dt_alias_exists() back to native_posix in practice). The test cases should hint the CI/CD platform to focus on native_posix (while not keeping anybody from running them on other platforms if required)~~
* ~~The RTC emulation device [is meant for testing purposes only](https://github.com/zephyrproject-rtos/zephyr/issues/19030#issuecomment-1615812406) but resides in the main driver tree which clutters devicetree and docs, may confuse users concerning its purpose (it confused me quite a bit), breaks encapsulation and creates redundant code. As customary, mock/fake drivers should be placed in the test tree.~~
* A few minor locking/memory issues can be fixed on-the-fly plus the fake driver can be simplified a bit (not worth mentioning details, see PR that follows).

UPDATE: Removed items due to review comments on the linked PR, see there. Won't fix.

**To Reproduce**
./scripts/twister -T tests/drivers/rtc -c --integration

**Expected behavior**

* use spinlocks instead of mutexes

**Impact**

minor impact:
* ~~missing test coverage - potential for regressions~~
* ~~some avoidable maintenance effort, unnecessary complexity~~
* ~~potential confusion of users re purpose of zephyr,rtc-emul~~ (true, but no longer part of this issue, see update above)
* ~~a few extra seconds of execution on the CI/CD pipeline plus unnecessarily long "skipped" reports~~
* unnecessary resource usage during tests (~~kernel resources, interrupts,~~ mutex)

**Logs and console output**

n/a

**Environment (please complete the following information):**

dev env, CI/CD

**Additional context**

See https://github.com/zephyrproject-rtos/zephyr/issues/19030#issuecomment-1615275932

### PR DESCRIPTION
**Pull request by ghost**
Fixes: #59901

See issue description for details.

### CODE REVIEW

**Comment by cfriedt** [comment]
We actually do want rtc_emul in drivers/rtc. The tests/ directory contains only unit and integration tests.

Additionally, the rtc_emul driver is able to provide the same basic service even if a platform does not have an actual hardware RTC.

This goes way back to some embedded boards where of the hw RTC was missing, it could be initialized to the **build date** for the firmware image before picking up e.g. NTP time.

**Comment by fgrandel** [comment]
> We actually do want rtc_emul in drivers/rtc. The tests/ directory contains only unit and integration tests.

Ok, fixed.

**Review by cfriedt — DISMISSED** [review]
LGTM - need @bjarki-trackunit to approve as well, ofc

Note: sorry - was trying to review each commit individually, but it really does not show up correctly via the Github UI. This is an approval only for the `[EXPERIMENTAL]` commit

**Review by cfriedt — COMMENTED** [review]
Note: this comment refers to the `LOCKED()` changes.

I see now why you added this - actually, it's fairly useful, but the issue is that we reach into the implementation with `!__i.key` and that should not happen in the public API (`zephyr/spinlock.h`, where callers should only see `k_spin_lock()` and `k_spin_unlock()`). Additionally, the private kernel side of things _does_ need to know the implementation detail, so the same macro should probably not be used for both the public side and private side.

Personally, I would say this change is not not a direct prerequisite for updating `rtc_emul.c` to use spinlocks, so I would suggest maybe dropping it here.

**Review by cfriedt — COMMENTED** [review]
I would just use `atomic_val_t` for both `counter` and `address`. With that, there is no need to cast.

Additionally, please fix the test to use `zassert_equal()` instead of `zassert_true()`.

Note: For the "tests: drivers: rtc: fix cast" commit

File: tests/drivers/rtc/rtc_api/testcase.yaml
Comment (cfriedt) [inline tests/drivers/rtc/rtc_api/testcase.yaml:31]: Actually, it would be easier if all of the `=y` options could just be added to `prj.conf`. I realize that there may be unique code paths where one option is enabled and another is disabled, but this creates a fair bit of code churn for CI, and so far, all drivers support ALARM and CALIBRATE.

File: drivers/rtc/rtc_emul.c
Comment (cfriedt) [inline drivers/rtc/rtc_emul.c:290]: Hmm... I see now why you added this. It's useful.

File: drivers/rtc/rtc_emul.c
Comment (cfriedt) [inline drivers/rtc/rtc_emul.c:250]: If possible, please remove things that are only formatting changes.

File: drivers/rtc/rtc_emul.c
Comment (cfriedt) [inline drivers/rtc/rtc_emul.c:254]: Same, please remove formatting changes.

File: drivers/rtc/rtc_emul.c
Comment (cfriedt) [inline drivers/rtc/rtc_emul.c:246]: I like this pattern but it would probably be better off as a separate tree-wide PR.

File: drivers/rtc/rtc_emul.c
Comment (cfriedt) [inline drivers/rtc/rtc_emul.c:268]: again, great pattern but it should go into a separate tree-wide PR. For now, please use the cumbersome manual approach with `k_spin_lock()`, `k_spin_unlock()`.

File: drivers/rtc/rtc_emul.c
Comment (cfriedt) [inline drivers/rtc/rtc_emul.c:299]: Technically no need for parenthesis, so it can just be `*timeptr`

File: drivers/rtc/rtc_emul.c
Comment (cfriedt) [inline drivers/rtc/rtc_emul.c:290]: Again - great pattern. Let's fork it out to a separate tree-wide PR and do it the old-fashioned way here to narrow the focus a bit.

File: drivers/rtc/rtc_emul.c
Comment (cfriedt) [inline drivers/rtc/rtc_emul.c:365]: Same - do it the old fashioned way, but please do either make an issue or a PR for the guarded macro.

File: drivers/rtc/rtc_emul.c
Comment (cfriedt) [inline drivers/rtc/rtc_emul.c:35]: The mutex->spinlock conversion is agreeably a much more lightweight synchronization primitive.

Just curious - what are we saving by switching from delayable work to timer? It isn't obvious to me if there is a major benefit at this time.

**Review by cfriedt — COMMENTED** [review]
Some comments. If possible, I think it would be better to narrow the focus a bit to simply changing from mutex to spinlock. However, I do like the spinlock macro that was introduced (high-level), but it would be nice if it were a tree-wide change, where appropriate.

**Comment by fgrandel** [comment]
> I do like the spinlock macro that was introduced (high-level), but it would be nice if it were a tree-wide change, where appropriate.

@cfriedt Did you see that this is a draft PR for that reason? The first commit introducing the macro does not belong to this PR, this PR will be rebased (or changed back to not using the macro) once #59884 has been merged or rejected. Is it ok, if I just ignore your comments re this pattern then? Or do I misunderstand you?

**Comment by cfriedt** [comment]
> @cfriedt Did you see that this is a draft PR for that reason? The first commit introducing the macro does not belong to this PR, this PR will be rebased (or changed back to not using the macro) once #59884 has been merged or rejected. Is it ok, if I just ignore your comments re this pattern then? Or do I misunderstand you?

Yes of course.

Would you prefer I comment here or on Discord?

**Comment by fgrandel** [comment]
> Would you prefer I comment here or on Discord?

To me both is equally fine. There seems to be some disagreement in the community depending on who you ask. Some see Discord critically because it's not searchable, others want to reduce verbosity in public issues. Choose according to your preferfence, I'm fine with any working style as long as it leads to a solution. :-)

File: tests/drivers/rtc/rtc_api/testcase.yaml
Comment (fgrandel) [reply tests/drivers/rtc/rtc_api/testcase.yaml:31]: Fixed.

File: drivers/rtc/rtc_emul.c
Comment (fgrandel) [reply drivers/rtc/rtc_emul.c:250]: @cfried These are actually introduced due to the additional code block around them:

before:

```
k_mutex_lock()
rtc_emul...()
k_mutex_unlock()
```
now

```
K_SPINLOCK_GUARDED_SECTION {
      rtc_emul...()
}
```

It's just difficult to see because of the preprocessor statements.

So I believe that in this case the formatting change is inevitable (as in the other newly introduced blocks). Do you agree?

File: drivers/rtc/rtc_emul.c
Comment (fgrandel) [reply drivers/rtc/rtc_emul.c:254]: Not ignored just resolved as a duplicate - will fix depending on the outcome of your previous comment.

File: drivers/rtc/rtc_emul.c
Comment (fgrandel) [reply drivers/rtc/rtc_emul.c:35]: 1) It's a simpler and IMO more obvious pattern (it removes some lines of code and does not need to be re-scheduled on every interrupt) - that's the main argument IMO
2) It considerably reduces the number of instructions on a time critical path (reducing latency and jitter of the generated tick signal)
3) It reduces the memory footprint of the data structure by 16 bytes IIRC (I looked at it through paholes to verify)

Do you think that justifies the change?

**Comment by fgrandel** [comment]
> I would just use `atomic_val_t` for both `counter` and `address`. With that, there is no need to cast.

@cfriedt I think there is no cast of counter any more. Probably github tricked you in looking at an old version of the commit? The only cast left is that of a pointer to atomic_val_t which is probably inevitable. Do you agree?

File: drivers/rtc/rtc_emul.c
Comment (fgrandel) [reply drivers/rtc/rtc_emul.c:246]: Ignored for now as not being part of this PR - see discussion below.

File: drivers/rtc/rtc_emul.c
Comment (fgrandel) [reply drivers/rtc/rtc_emul.c:268]: Ignored for now as not being part of this PR - see discussion below.

File: drivers/rtc/rtc_emul.c
Comment (fgrandel) [reply drivers/rtc/rtc_emul.c:290]: Ignored for now as not being part of this PR - see discussion below.

File: drivers/rtc/rtc_emul.c
Comment (fgrandel) [reply drivers/rtc/rtc_emul.c:365]: Ignored for now as not being part of this PR - see discussion below.

**Comment by fgrandel** [comment]
> Additionally, please fix the test to use `zassert_equal()` instead of `zassert_true()`.

Fixed.

File: drivers/rtc/Kconfig
Comment (bjarki-trackunit) [inline drivers/rtc/Kconfig:5]: Device driver APIs are not experimental features, they follow the lifecycle described [here](https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#api-lifecycle) The RTC device driver API is at this time `unstable` by the actual descriptions used for device driver APIs. This is therefore an inappropriate change.

File: drivers/rtc/rtc_emul.c
Comment (bjarki-trackunit) [inline drivers/rtc/rtc_emul.c:13]: This change us better achieved by using the CONTAINER_OF macro, which removes the need for this structure, while keeping the work queue item.

File: drivers/rtc/rtc_emul.c
Comment (bjarki-trackunit) [inline drivers/rtc/rtc_emul.c:35]: Usage of the work queue item is preferred to a timer since work, like invoking a callback, shall not be performed from within a timer callback. Work must be delegated to a thread, like the work queue. This is the reason ISR callbacks delegate work to be performed to the work queue.

Keep in mind that this driver is not time critical, it is not a real piece of hardware meant to actually be used as a functional RTC, its an emulator, which provides similar behavior to an RTC, allowing for rapid application development if real hardware is not available.

File: include/zephyr/spinlock.h
Comment (bjarki-trackunit) [inline include/zephyr/spinlock.h:32]: This change must be placed in its own PR.

File: kernel/include/kernel_internal.h
Comment (bjarki-trackunit) [inline kernel/include/kernel_internal.h:230]: This should also be moved out of this PR

File: tests/drivers/rtc/rtc_api/prj.conf
Comment (bjarki-trackunit) [inline tests/drivers/rtc/rtc_api/prj.conf:5]: The RTC API only tests the mandatory RTC features by default. Additional features like the alarm might not be implemented, which will lead to failed tests for those devices. This change shall be undone for that reason.

File: tests/drivers/rtc/rtc_api/src/test_update_callback.c
Comment (bjarki-trackunit) [inline tests/drivers/rtc/rtc_api/src/test_update_callback.c:28]: This value is not atomic, its storing the value of an atomic. This change introduces a bug if the atomic_val_t is not a uint32_t, as you typecast it on line 54.

File: tests/drivers/rtc/rtc_api/testcase.yaml
Comment (bjarki-trackunit) [inline tests/drivers/rtc/rtc_api/testcase.yaml:12]: native_posix is not an integration platform, this test suite is built to run on any platform that has an RTC, real and emulated like native_posix or qemu [see](https://docs.zephyrproject.org/latest/develop/test/twister.html)

File: tests/drivers/rtc/rtc_api_helpers/testcase.yaml
Comment (bjarki-trackunit) [inline tests/drivers/rtc/rtc_api_helpers/testcase.yaml:11]: See previous comment

**Review by bjarki-andreasen — CHANGES_REQUESTED** [review]
I agree with changing the rtc_emul driver to use a spinlock instead of a mutex as @cfriedt has previously noted as well.

Changing the zephyr test stuite to use zassert_equal instead of zassert_true I am apathetic to.

Intoducing a change to the spinlock API is a big change that requires its own PR, I have no issue with that change in general, but I lack the knowledge required to review that change.

As my comments reveal, the remaining changes are either incorrect or IMO unwarranted.

File: drivers/rtc/rtc_emul.c
Comment (fgrandel) [reply drivers/rtc/rtc_emul.c:35]: Obsolete/removed - see @bjarki-trackunit 's comment below.

File: drivers/rtc/rtc_emul.c
Comment (fgrandel) [reply drivers/rtc/rtc_emul.c:13]: Obsolete: change was reverted.

File: drivers/rtc/rtc_emul.c
Comment (fgrandel) [reply drivers/rtc/rtc_emul.c:35]: Timer conversion is now removed from change set.

File: include/zephyr/spinlock.h
Comment (fgrandel) [reply include/zephyr/spinlock.h:32]: See PR description - this change already is in a separate change set. That's why the PR is a draft. Depending on the outcome of the other PR this PR will either stay as it is or be changed to use the classical spinlock pattern.

File: kernel/include/kernel_internal.h
Comment (fgrandel) [reply kernel/include/kernel_internal.h:230]: Duplicate.

File: tests/drivers/rtc/rtc_api/prj.conf
Comment (fgrandel) [reply tests/drivers/rtc/rtc_api/prj.conf:5]: The change was reverted/removed from the change set.

File: tests/drivers/rtc/rtc_api/src/test_update_callback.c
Comment (fgrandel) [reply tests/drivers/rtc/rtc_api/src/test_update_callback.c:28]: The change was removed from the change set.

File: tests/drivers/rtc/rtc_api/testcase.yaml
Comment (fgrandel) [reply tests/drivers/rtc/rtc_api/testcase.yaml:12]: The change was removed from the change set.

File: tests/drivers/rtc/rtc_api_helpers/testcase.yaml
Comment (fgrandel) [reply tests/drivers/rtc/rtc_api_helpers/testcase.yaml:11]: Duplicate.

File: drivers/rtc/Kconfig
Comment (cfriedt) [reply drivers/rtc/Kconfig:5]: Yeah, I wouldn't change the spinlock API in this pr. If the api is experimental or not, it's kind of irrelevant.

Comment (fgrandel) [reply drivers/rtc/Kconfig:5]: @cfriedt That spinlock stuff is now merged in main, so this PR is no longer a draft. The question I asked above wasn't so much about the spinlock (which I never intended to be part of this PR) but about marking the RTC driver API experimental. I'm a bit confused, because it is shown as experimental in the [API overview](https://docs.zephyrproject.org/latest/develop/api/overview.html) plus experimental API drivers are usually marked as such in Kconfig to avoid confusion and (more importantly) flagged as `select EXPERIMENTAL` so they produce automated warnings when used which you confirmed in your first review. @bjarki-trackunit  disagreed on that point, that's why I asked you both to quickly co-ordinate, so that I know what the final consensus will be on that point and finalize this PR then.

File: tests/drivers/rtc/rtc_api/testcase.yaml
Comment (fgrandel) [reply tests/drivers/rtc/rtc_api/testcase.yaml]: Reverted due to feedback by @bjarki-trackunit - see below.

File: drivers/rtc/rtc_emul.c
Comment (fgrandel) [reply drivers/rtc/rtc_emul.c]: Fixed in a separate commit.

File: drivers/rtc/Kconfig
Comment (fgrandel) [reply drivers/rtc/Kconfig]: Clarified with cfried that this change will be removed. The corresponding ticket was closed as "won't fix".

**Review by cfriedt — DISMISSED** [review]
Just a mention, maybe for @bjarki-trackunit to follow up, but `zassert_equal(x, 0, ..)` can be replaced with `zassert_ok(x, ...)`

**Review by bjarki-andreasen — CHANGES_REQUESTED** [review]
Looks good, however, as @cfriedt discovered, the macro `zassert_ok` is safer than `zassert_equal` where applicable. I will request you use that macro instead where applicable. I will sanity check the changes locally and approve the PR once its ready :)

**Comment by fgrandel** [comment]
> the macro `zassert_ok` is safer than `zassert_equal` 

@bjarki-trackunit  Fixed.

**Review by cfriedt — APPROVED** [review]

**Review by bjarki-andreasen — APPROVED** [review]
Emulated RTC driver and RTC API test suite behavior validated. Nice work :)

**Pull request merged**
